### PR TITLE
Refactoring - M3 feedback

### PR DESCRIPTION
**Pull request by Athanaze**
The aim of this PR is to address the issues observed during the grading of M3.

### CODE REVIEW

File: app/src/main/java/com/monkeyteam/chimpagne/ui/HomeScreen.kt [hunk 1]
```diff
@@ -67,22 +67,6 @@ import com.monkeyteam.chimpagne.ui.utilities.PromptLogin
...
 import com.monkeyteam.chimpagne.viewmodels.FindEventsViewModel
 
-class LocationViewModel(myContext: Context) {
-  private val fusedLocationClient = LocationServices.getFusedLocationProviderClient(myContext)
-
-  fun startLocationUpdates(
-      myContext: Context,
-      onLocationSuccess: (lat: Double, lng: Double) -> Unit
-  ) {
-    if (ActivityCompat.checkSelfPermission(myContext, Manifest.permission.ACCESS_FINE_LOCATION) ==
```
Comment (arnaudlvq) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/HomeScreen.kt:77]: ACCESS_FINE_LOCATION should be asked with ACCESS_COARSE_LOCATION as discibed in this documentation page :

[here](https://developer.android.com/develop/sensors-and-location/location/permissions?hl=fr#foreground)

File: app/src/main/java/com/monkeyteam/chimpagne/ui/HomeScreen.kt [hunk 2]
```diff
@@ -67,22 +67,6 @@ import com.monkeyteam.chimpagne.ui.utilities.PromptLogin
...
-        PackageManager.PERMISSION_GRANTED) {
-      fusedLocationClient.lastLocation.addOnSuccessListener { location ->
-        location?.let { onLocationSuccess(it.latitude, it.longitude) }
-      }
-    }
-  }
-}
-
 @SuppressLint("StateFlowValueCalledInComposition", "MissingPermission")
```
Comment (arnaudlvq) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/HomeScreen.kt:70]: Missing permissions ? for the feed ? You might explain why you include this @SuppressLint

File: app/src/main/java/com/monkeyteam/chimpagne/ui/HomeScreen.kt [hunk 3]
```diff
@@ -135,7 +103,7 @@ fun HomeScreen(
           findViewModel.uiState.value.events.forEach { (_, u) -> eventsNearMe.add(u) }
 
           findViewModel.uiState.value.selectedLocation?.let {
-            closestEventsState.value = getClosestNEvent(eventsNearMe, N_CLOSEST, it)
+            closestEventsState.value = getClosestNEvent(eventsNearMe, it)
```
Comment (arnaudlvq) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/HomeScreen.kt:106]: Perfect, 4 was too little

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/polls/CreatePollDialog.kt [hunk 4]
```diff
@@ -58,7 +58,7 @@ fun CreatePollDialog(
                   text = stringResource(id = R.string.chimpagne_confirm),
                   modifier = Modifier.testTag("confirm poll button"),
                   onClick = {
-                    if (title == "" || query == "" || options.contains("")) {
+                    if (title.isEmpty() || query.isEmpty() || options.contains("")) {
```
Comment (arnaudlvq) [inline app/src/main/java/com/monkeyteam/chimpagne/ui/event/polls/CreatePollDialog.kt:61]: I like the refactoring you do, they are appreciated as much as in your reviews B)

**Review by arnaudlvq — COMMENTED** [review]
### Just a little bit more ?

I'll just leave a comment this time, I'm not sure any version of the application would be accepted without the COARSE location authorization request.

The documentation you've provided is excellent, but the definition of the parameters is missing here and there

**Review by Kuriv2001 — APPROVED** [review]
Looks good! Some higher test coverage would be appreciated

File: app/src/main/java/com/monkeyteam/chimpagne/model/feed/FeedAlgorithm.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package com.monkeyteam.chimpagne.model.feed
+
+import com.monkeyteam.chimpagne.model.database.ChimpagneEvent
+import com.monkeyteam.chimpagne.model.location.Location
+
+val N_CLOSEST = 4
```
Comment (QuicheLorraine3600) [inline app/src/main/java/com/monkeyteam/chimpagne/model/feed/FeedAlgorithm.kt:6]: This value is ridiculously low, why do we need to take only the N_CLOSEST events ?

File: app/src/main/java/com/monkeyteam/chimpagne/model/feed/FeedAlgorithm.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+fun getClosestNEvent(
+    li: List<ChimpagneEvent>,
+    myLocation: Location,
+): List<ChimpagneEvent> {
+  if (li.size <= N_CLOSEST) return li
+
+  val sortedEvents =
```
Comment (QuicheLorraine3600) [inline app/src/main/java/com/monkeyteam/chimpagne/model/feed/FeedAlgorithm.kt:14]: You are sorting the whole event list so this doesn't improve performances. 
Just returning the event list sorted by distance is enough

File: app/src/main/java/com/monkeyteam/chimpagne/model/location/LocationViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package com.monkeyteam.chimpagne.model.location
+
+import android.Manifest
+import android.content.Context
+import android.content.pm.PackageManager
+import androidx.core.app.ActivityCompat
+import com.google.android.gms.location.LocationServices
+
+class LocationViewModel(myContext: Context) {
```
Comment (QuicheLorraine3600) [inline app/src/main/java/com/monkeyteam/chimpagne/model/location/LocationViewModel.kt:9]: Why do you call this a view model if it doesn't store any state and doesn't even extends the ViewModel class ?

File: app/src/main/java/com/monkeyteam/chimpagne/ui/event/polls/CreatePollDialog.kt (hunk 4 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/ui/event/polls/CreatePollDialog.kt:61]: still unsure of the usefulness of this

File: app/src/main/java/com/monkeyteam/chimpagne/viewmodels/EventViewModel.kt [hunk 8]
```diff
@@ -243,15 +242,7 @@ class EventViewModel(
...
-  fun joinEvent(
-      eventId: ChimpagneEventId = _uiState.value.id,
-      onSuccess: () -> Unit = {},
-      onFailure: (Exception) -> Unit = {}
-  ) {
-    joinTheEvent(onSuccess, onFailure)
-  }
-
-  fun joinTheEvent(onSuccess: () -> Unit = {}, onFailure: (Exception) -> Unit = {}) {
+  fun joinEvent(onSuccess: () -> Unit = {}, onFailure: (Exception) -> Unit = {}) {
```
Comment (QuicheLorraine3600) [inline app/src/main/java/com/monkeyteam/chimpagne/viewmodels/EventViewModel.kt:245]: thank you, it was really disturbing to have joinEvent and joinTheEvent in the same class for the same thing

**Review by QuicheLorraine3600 — DISMISSED** [review]
Nice job overall. However, the feed still needs improvements to be useful. Taking only the 4 closest events isn't enough. As we take more events for the feed, I would like to change my position related to the filters used by the feed: events I'm already in should be displayed on it.

File: app/src/main/java/com/monkeyteam/chimpagne/model/location/LocationViewModel.kt (hunk 7 above)
Comment (Athanaze) [reply app/src/main/java/com/monkeyteam/chimpagne/model/location/LocationViewModel.kt:9]: It does store the state of fusedLocationClient

File: app/src/main/java/com/monkeyteam/chimpagne/model/feed/FeedAlgorithm.kt (hunk 5 above)
Comment (Athanaze) [reply app/src/main/java/com/monkeyteam/chimpagne/model/feed/FeedAlgorithm.kt:6]: I mean 1000000 events would be too much for example.
We have to take a finite amount. Changing the algorithm is not part of this pr.

File: app/src/main/java/com/monkeyteam/chimpagne/model/location/LocationViewModel.kt (hunk 7 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/model/location/LocationViewModel.kt:9]: it's still not a view model, you do not extend the VIewModel class

File: app/src/main/java/com/monkeyteam/chimpagne/model/feed/FeedAlgorithm.kt (hunk 5 above)
Comment (QuicheLorraine3600) [reply app/src/main/java/com/monkeyteam/chimpagne/model/feed/FeedAlgorithm.kt:6]: Why would it be too much ? From a performance perspective, you're already sorting those 1000000 events locally. LazyColumn is made to display a huge amount of items.

**Review by arnaudlvq — APPROVED** [review]
Thank you for changing the feed size from 4 to 50 as discussed this day

**Pull request merged**